doc/developer/design/20220228_kafka_headers.md

```diff
@@ -0,0 +1,87 @@
+# Kafka Headers
+
+## Summary
+
+Support the `headers` section of kafka messages in sources. This is a `string` -> `bytes` multimap of arbitrary
+data. The purpose of this design document is to describe how we can implement this.
+
+## Goals
+
+Add basic support for unstructured data from headers that can be cast/converted later
+in the sql layer
+
+## Non-Goals
+
+It is a non-goal to support
+[Kafka Connect `Header`'s](https://kafka.apache.org/20/javadoc/index.html?org/apache/kafka/connect/header/Header.html),
+which are, as far as I can tell, _dynamically-schema'd_ headers built on kafka headers. This is similar to the
+dynamic schema's supported by Kafka Connect.
+
+
+## Description
+
+Add an `INCLUDE HEADERS` syntax that allows users to declare that they want a `headers` column for their source,
+as provided in kafka sources. This will provide a list of from (text, bytes (`bytea`)) pairs. This is because
+Kafka headers can have multiple values per-key, and preserve order.
+Users are expected to cast these bytes values to the types they want in a later view.
+This option will only be available for `ENVELOPE NONE` and `ENVELOPE UPSERT`.
+
+## Proposed syntax changes
+
+Kafka source creation will be extended so that `INCLUDE` has a new `HEADERS` option, so the full
+syntax becomes:
+
+```
+INCLUDE (
+    KEY (AS <name>)? |
+    TIMESTAMP (AS <name>)? |
+    OFFSET (AS <name>)? |
+    PARTITION (AS <name>)? |
+    TOPIC (AS <name>)? |
+    HEADERS (AS <name>)?
+)+
+```
+as an example:
+```sql
+CREATE MATERIALIZED SOURCE avroavro
+FROM KAFKA BROKER '...' TOPI '...`
+FORMAT ...
+INCLUDE HEADERS
+ENVELOPE ...
+```
+
+This will add a new column to the row from the kafka source that is of type
+`list[Record { name: text, values: bytea}]` with the name `headers`, with
```

Comment:
It seems like Kafka consistently refers to the header name as "key", so I think this record type should do the same! Also, looks like you can mark it non-nullable.